File: Drum/sticking-engine/sticking_engine/engine.py
"""
engine.py

RULE 27 — UNIFIED PERFORMANCE DATA MODEL, EVENT GRAPH & DATA CONTRACT
  (implemented in datamodel.py; this file is the orchestrator that produces
  the single UnifiedPerformance / FinalValidatedPerformance object)

RULE 28 — ENGINE MODULE ARCHITECTURE & RULE ORCHESTRATION
  Principle: "Orchestrator owns when; specialized modules own what/how;
  circular hidden dependencies are forbidden." -> `Engine.run()` below is
  the ONLY place that decides call order; every rule module it calls is a
  pure function or a small explicit class with no import back into engine.py.

RULE 36 — END-TO-END PERFORMANCE SOLVER & FINAL DRUMMER BEHAVIOR INTEGRATION
  End-to-End Loop: "Perceive -> Understand -> Remember -> Predict -> Generate
  -> Simulate -> Decide -> Move -> Impact -> Recover -> Adapt -> Continue."
  `Engine.run()` follows exactly this order.

RULE 37 — ENGINE IMPLEMENTATION SPECIFICATION, CLASS MODEL & INTER-MODULE
INTERFACES
  Ownership Rule: "One decision, one owner, one source of truth."
    - Rule 7/8/11/29 solver owns HAND assignment.
    - Rule 9 owns TECHNIQUE.
    - Rule 10/13-19 own MOTION/RECOVERY.
    - Rule 25 owns VALIDATION/APPROVAL.
  Engine never re-derives a decision another module already owns.

RULE 38 — CLASS, INTERFACE & DATA SCHEMA SPECIFICATION
  (implemented in datamodel.py: typed dataclasses, explicit units, explicit
  Candidate/Decision/Plan/Result/Context/State distinctions.)

RULE 40 — SOLVER PERFORMANCE, CACHING, PARALLELISM & REAL-TIME OPTIMIZATION
  Optimization Rule: "Cache stable work, reject impossible candidates early,
  parallelize only read-only/speculative work ... keep correctness invariant
  across FAST/HIGH modes." -> `mode="FAST"` uses a narrower beam/window;
  `mode="HIGH"` widens both. Both paths run the exact same functions, so
  results only differ in search thoroughness, never in code path.
"""

from __future__ import annotations
from typing import List, Optional, Dict
from dataclasses import dataclass

from .datamodel import (
    Instrument, Limb, DrumEvent, DrummerState, DrummerStyleProfile,
    PerformanceIntentContext, FatigueState, PerformanceEvent, new_id,
    FinalValidatedPerformance,
)
from .rule01_input import parse_midi_file, from_note_list
from .rule02_mapping import DrumMappingProfile, map_events
from .rule03_04_timing_density import analyze_timing, analyze_density
from .rule05_pattern import classify_patterns, role_contexts
from .rule06_reachability import neutral_position
from .rule08_11_29_39_solver import solve_sticking
from .rule09_technique import select_technique
from .rule10_recovery import plan_recovery
from .rule12_humanization import humanize_timing, humanize_velocity
from .rule13_19_24_motion import (
    build_motion_plan, apply_rig_envelope, compute_impact,
    append_recovery_keyframe, compute_body_state, assemble_timeline,
    check_audio_animation_sync,
)
from .rule20_31_33_memory import PerformanceMemory
from .rule21_23_32_35_profile import (
    update_fatigue, ProfileCalibrator, build_effective_style, make_idiom_context,
)
from .rule25_validation import validate_and_repair
from .rule30_state_machine import DrummerStateMachine


# ---------------------------------------------------------------------------
# RULE 40 — search-effort presets (same code path, different thoroughness)
# ---------------------------------------------------------------------------
SEARCH_PRESETS = {
    "FAST": {"window_size": 6, "beam_width": 3},
    "HIGH": {"window_size": 14, "beam_width": 8},
}


@dataclass
class EngineConfig:
    """RULE 37/38: the engine's one explicit configuration surface. Nothing
    inside Engine.run() reads ambient/global config — everything flows
    through this object."""
    style: DrummerStyleProfile
    intent: PerformanceIntentContext
    genre: str = "generic"
    seed: int = 42
    mode: str = "HIGH"          # "FAST" | "HIGH" (Rule 40)
    drum_channel: Optional[int] = None
    mapping_profile: Optional[DrumMappingProfile] = None


class Engine:
    """RULE 28 orchestrator. Owns *when* each rule module runs; owns nothing
    about *what* each decision is."""

    def __init__(self, config: Optional[EngineConfig] = None):
        self.config = config or EngineConfig(
            style=DrummerStyleProfile(), intent=PerformanceIntentContext()
        )
        self.memory = PerformanceMemory()
        self.calibrator = ProfileCalibrator()

    # -- RULE 36 End-to-End Loop -------------------------------------------
    def run(self, midi_path: Optional[str] = None,
            note_list: Optional[List[dict]] = None) -> FinalValidatedPerformance:
        cfg = self.config
        preset = SEARCH_PRESETS.get(cfg.mode, SEARCH_PRESETS["HIGH"])

        # ---- Perceive (Rule 1) ----
        if midi_path is not None:
            normalized = parse_midi_file(midi_path, drum_channel=cfg.drum_channel)
        elif note_list is not None:
            normalized = from_note_list(note_list)
        else:
            raise ValueError("Engine.run requires midi_path or note_list")

        # ---- Understand (Rules 2-5) ----
        mapping_profile = cfg.mapping_profile or DrumMappingProfile()
        drum_events: List[DrumEvent] = map_events(normalized.events, mapping_profile)
        drum_events = [e for e in drum_events if e.is_playable]
        drum_events.sort(key=lambda e: e.time_seconds)

        timing_ctx = analyze_timing(drum_events, normalized.tempo_map, normalized.time_signature_map)
        density_ctx = analyze_density(drum_events)
        pattern_ctx = classify_patterns(drum_events, timing_ctx, density_ctx)
        # RULE 5 -> RULE 7/8. What each note is FOR is an input to the
        # sticking, not a report about it: it is what tells the solver that
        # a run of hi-hat notes is one hand keeping time rather than a
        # series of unrelated notes to alternate across.
        roles = role_contexts(pattern_ctx)

        # ---- Remember / Predict (Rules 20/31/32/35 setup) ----
        idiom = make_idiom_context(cfg.genre)
        effective_style = build_effective_style(
            cfg.style, self.calibrator.learned, cfg.intent, FatigueState(), idiom
        )

        state = DrummerState(style=effective_style, intent=cfg.intent)
        for limb, ls in state.limbs.items():
            ls.position = neutral_position(limb)
            ls.last_action_time_s = -999.0
            ls.ready_time_s = -999.0

        sm = DrummerStateMachine(state)

        # ---- Generate / Simulate / Decide (Rules 6-12, 29, 39) ----
        # RULE 11: all four limbs in ONE solve. Splitting the hands from the
        # feet and concatenating afterwards would mean the hands choose
        # without knowing a kick lands on the same beat, which is precisely
        # what Rule 11's Core Principle rules out.
        all_decisions = solve_sticking(
            drum_events, sm.state,
            window_size=preset["window_size"], beam_width=preset["beam_width"],
            roles=roles,
        )

        self.memory.record_committed(all_decisions)
        decision_by_event = {d.event_id: d for d in all_decisions}
        event_by_id: Dict[str, DrumEvent] = {e.event_id: e for e in drum_events}

        # ---- Move / Impact / Recover (Rules 9-10, 13-19, 23-24) ----
        # Deterministic time-ordered replay: the solver already fixed WHICH
        # limb plays WHAT; this pass owns HOW (technique/motion), touching
        # nothing the solver decided (Rule 37 Ownership Rule).
        replay_position = {l: neutral_position(l) for l in Limb}
        replay_ready_time = {l: -999.0 for l in Limb}
        replay_last_time = {l: None for l in Limb}
        fatigue = FatigueState()

        ordered = sorted(all_decisions, key=lambda d: event_by_id[d.event_id].time_seconds)

        performance_events: List[PerformanceEvent] = []
        motion_plans = []
        technique_by_event = {}
        body_states = []

        for decision in ordered:
            ev = event_by_id[decision.event_id]
            limb = decision.limb

            technique = select_technique(
                ev, limb, effective_style, replay_last_time[limb]
            )
            technique_by_event[ev.event_id] = technique

            recovery = plan_recovery(ev.time_seconds, limb, technique)

            human_timing = humanize_timing(
                ev.event_id, ev.time_seconds, effective_style, cfg.seed
            )
            humanized_velocity = humanize_velocity(
                ev.event_id, ev.velocity, effective_style, cfg.seed
            )

            target_pos = (ev.target.x, ev.target.y, ev.target.height)
            plan = build_motion_plan(
                ev.event_id, limb, technique, replay_position[limb], target_pos,
                human_timing, replay_ready_time[limb],
            )
            plan = apply_rig_envelope(plan)
            plan = append_recovery_keyframe(plan, recovery)
            motion_plans.append(plan)

            impact = compute_impact(plan, ev.instrument, technique)

            fatigue = update_fatigue(fatigue, limb, technique, ev.time_seconds)

            replay_position[limb] = (ev.target.x, ev.target.y, ev.target.height + recovery.rebound_height_m)
            replay_ready_time[limb] = recovery.ready_time_s
            replay_last_time[limb] = ev.time_seconds

            active_positions = {l: p for l, p in replay_position.items()}
            body_states.append(compute_body_state(ev.time_seconds, active_positions))

            is_dominant_hand = (
                limb in (Limb.RIGHT_HAND, Limb.LEFT_HAND)
                and limb.value == (effective_style.dominant_hand.value + "H")
            )
            # NOTE: body-crossing is scored inside Rule 7's candidate search
            # (StickingCandidate.crosses_body) but a committed SequenceDecision
            # does not carry that tag forward; `crossed=False` is a safe
            # default here and the interface leaves room to thread the real
            # flag through SequenceDecision in a future revision.
            self.calibrator.observe(chosen_dominant=is_dominant_hand, crossed=False)

            performance_events.append(PerformanceEvent(
                event_id=ev.event_id,
                source_id=ev.source_id,
                time_seconds=human_timing.performed_time_s,
                limb=limb,
                instrument=ev.instrument,
                stroke_type=technique.stroke_type,
                velocity=humanized_velocity,
                microtiming_offset_ms=human_timing.microtiming_offset_ms,
                dynamic_level=technique.dynamic_level,
                rule_trace=["rule08/29", "rule09", "rule10", "rule12", "rule13-19"],
            ))

        # ---- Assemble timeline (Rule 19) & sync (Rule 24) ----
        timeline = assemble_timeline(motion_plans, technique_by_event)
        check_audio_animation_sync(timeline)

        # ---- Validate & repair (Rule 25) ----
        validation = validate_and_repair(performance_events)

        # ---- Continue (Rule 26 will pick this up at playback time) ----
        return FinalValidatedPerformance(
            events=performance_events,
            animation=timeline,
            validation=validation,
            duration_s=timeline.duration_s,
            seed=cfg.seed,
        )

File: Drum/sticking-engine-ts/src/engine.ts
/**
 * engine.ts — the orchestrator.
 *
 * RULE 28 — ENGINE MODULE ARCHITECTURE: "orchestrator owns WHEN;
 *   specialized modules own WHAT/HOW; circular hidden dependencies are
 *   forbidden." `run()` is the ONLY place that decides call order, and
 *   no rule module imports back into this file.
 * RULE 36 — END-TO-END SOLVER: "Perceive -> Understand -> Remember ->
 *   Predict -> Generate -> Simulate -> Decide -> Move -> Impact ->
 *   Recover -> Adapt -> Continue." `run()` follows exactly that order.
 * RULE 37 — "One decision, one owner, one source of truth." The solver
 *   owns HAND, Rule 9 owns TECHNIQUE, Rules 10/13–19 own MOTION, Rule 25
 *   owns APPROVAL. The engine never re-derives a decision another module
 *   already owns.
 * RULE 40 — FAST/HIGH are the same code path with a narrower or wider
 *   beam and window, so results differ in thoroughness, never in logic.
 */

import type {
  DrumEvent,
  DrummerStyleProfile,
  FinalValidatedPerformance,
  Limb,
  MotionPlan,
  PerformanceEvent,
  PerformanceIntentContext,
  Point3,
  SequenceDecision,
  TechniquePlan,
} from './datamodel.js';
import { LIMBS, drummerState, drummerStyle, fatigueState, performanceIntent } from './datamodel.js';
import { fromNoteList, type NoteListEntry } from './rule01-input.js';
import { drumMappingProfile, mapEvents, type DrumMappingProfile } from './rule02-mapping.js';
import { analyzeDensity, analyzeTiming } from './rule03-04-timing-density.js';
import { classifyPatterns } from './rule05-pattern.js';
import { neutralPosition } from './rule06-reachability.js';
import { assignFeet, solveSticking } from './rule08-11-29-39-solver.js';
import { selectTechnique } from './rule09-technique.js';
import { planRecovery } from './rule10-recovery.js';
import { humanizeTiming, humanizeVelocity } from './rule12-humanization.js';
import {
  appendRecoveryKeyframe,
  applyRigEnvelope,
  assembleTimeline,
  checkAudioAnimationSync,
  computeBodyState,
  computeImpact,
} from './rule13-19-24-motion.js';
import { PerformanceMemory } from './rule20-31-33-memory.js';
import {
  ProfileCalibrator,
  buildEffectiveStyle,
  makeIdiomContext,
  updateFatigue,
} from './rule21-23-32-35-profile.js';
import { validateAndRepair } from './rule25-validation.js';
import { DrummerStateMachine } from './rule30-state-machine.js';

/** RULE 40 — search-effort presets. Same code path, different thoroughness. */
export const SEARCH_PRESETS: Readonly<Record<string, { windowSize: number; beamWidth: number }>> = {
  FAST: { windowSize: 6, beamWidth: 3 },
  HIGH: { windowSize: 14, beamWidth: 8 },
};

/**
 * The engine's one explicit configuration surface. Nothing inside
 * `run()` reads ambient or global config -- everything flows through here.
 */
export interface EngineConfig {
  style: DrummerStyleProfile;
  intent: PerformanceIntentContext;
  genre: string;
  seed: number;
  /** "FAST" | "HIGH" (Rule 40). */
  mode: string;
  mappingProfile?: DrumMappingProfile;
}

export function engineConfig(overrides: Partial<EngineConfig> = {}): EngineConfig {
  return {
    style: overrides.style ?? drummerStyle(),
    intent: overrides.intent ?? performanceIntent(),
    genre: overrides.genre ?? 'generic',
    seed: overrides.seed ?? 42,
    mode: overrides.mode ?? 'HIGH',
    ...(overrides.mappingProfile !== undefined ? { mappingProfile: overrides.mappingProfile } : {}),
  };
}

export class Engine {
  readonly config: EngineConfig;
  readonly memory = new PerformanceMemory();
  readonly calibrator = new ProfileCalibrator();

  constructor(config?: Partial<EngineConfig>) {
    this.config = engineConfig(config ?? {});
  }

  /** RULE 36's end-to-end loop. */
  run(input: { noteList: readonly NoteListEntry[] }): FinalValidatedPerformance {
    const cfg = this.config;
    // An unrecognized mode falls back to HIGH rather than failing --
    // Rule 40's own behaviour, and the two modes differ only in search
    // thoroughness, so the fallback is never the wrong ANSWER.
    const preset = SEARCH_PRESETS[cfg.mode] ?? SEARCH_PRESETS['HIGH'];
    if (preset === undefined) throw new Error('SEARCH_PRESETS has lost its HIGH preset');

    // ---- Perceive (Rule 1) ----
    const normalized = fromNoteList(input.noteList);

    // ---- Understand (Rules 2-5) ----
    const mappingProfile = cfg.mappingProfile ?? drumMappingProfile();
    let drumEvents: DrumEvent[] = mapEvents(normalized.events, mappingProfile);
    drumEvents = drumEvents.filter((e) => e.isPlayable);
    drumEvents.sort((a, b) => a.timeSeconds - b.timeSeconds);

    const timing = analyzeTiming(drumEvents, normalized.tempoMap, normalized.timeSignatureMap);
    const density = analyzeDensity(drumEvents);
    // Timing/density/pattern context is available to any rule that wants
    // it; the reference solver uses reachability and style scoring
    // directly, which already reacts to density through available-time
    // windows. Computed anyway because the Python computes it, and
    // because Rule 5 consumes ID-counter values -- skipping it would
    // shift every later ID and make a parity diff unreadable.
    classifyPatterns(drumEvents, timing, density);

    // ---- Remember / Predict (Rules 20/31/32/35 setup) ----
    const idiom = makeIdiomContext(cfg.genre);
    const effectiveStyle = buildEffectiveStyle(
      cfg.style,
      this.calibrator.learned,
      cfg.intent,
      fatigueState(),
      idiom,
    );

    const state = drummerState({ style: effectiveStyle, intent: cfg.intent });
    for (const limb of LIMBS) {
      const ls = state.limbs[limb];
      ls.position = neutralPosition(limb);
      ls.lastActionTimeS = -999.0;
      ls.readyTimeS = -999.0;
    }
    const sm = new DrummerStateMachine(state);

    // ---- Generate / Simulate / Decide (Rules 6-12, 29, 39) ----
    const manualEvents = drumEvents.filter((e) => !e.target.isFootTarget);
    const footEvents = drumEvents.filter((e) => e.target.isFootTarget);

    const stickingDecisions = solveSticking(
      manualEvents,
      sm.state,
      preset.windowSize,
      preset.beamWidth,
    );
    const footDecisions = assignFeet(footEvents, sm.state);
    const allDecisions: SequenceDecision[] = [...stickingDecisions, ...footDecisions];

    this.memory.recordCommitted(allDecisions);
    const eventById = new Map(drumEvents.map((e) => [e.eventId, e]));

    // ---- Move / Impact / Recover (Rules 9-10, 13-19, 23-24) ----
    // A deterministic time-ordered replay. The solver already fixed
    // WHICH limb plays WHAT; this pass owns HOW, and touches nothing the
    // solver decided (Rule 37's ownership rule).
    const replayPosition = {} as Record<Limb, Point3>;
    const replayReadyTime = {} as Record<Limb, number>;
    const replayLastTime = {} as Record<Limb, number | undefined>;
    for (const limb of LIMBS) {
      replayPosition[limb] = neutralPosition(limb);
      replayReadyTime[limb] = -999.0;
      replayLastTime[limb] = undefined;
    }
    let fatigue = fatigueState();

    const ordered = [...allDecisions].sort(
      (a, b) =>
        (eventById.get(a.eventId) as DrumEvent).timeSeconds -
        (eventById.get(b.eventId) as DrumEvent).timeSeconds,
    );

    const performanceEvents: PerformanceEvent[] = [];
    const motionPlans: MotionPlan[] = [];
    const techniqueByEvent = new Map<string, TechniquePlan>();

    for (const decision of ordered) {
      const ev = eventById.get(decision.eventId) as DrumEvent;
      const limb = decision.limb;

      const technique = selectTechnique(ev, limb, effectiveStyle, replayLastTime[limb]);
      techniqueByEvent.set(ev.eventId, technique);

      const recovery = planRecovery(ev.timeSeconds, limb, technique);

      const humanTiming = humanizeTiming(ev.eventId, ev.timeSeconds, effectiveStyle, cfg.seed);
      const humanizedVelocity = humanizeVelocity(ev.eventId, ev.velocity, effectiveStyle, cfg.seed);

      const targetPos: Point3 = [ev.target.x, ev.target.y, ev.target.height];
      let plan = buildMotionPlanFor(
        ev.eventId,
        limb,
        technique,
        replayPosition[limb],
        targetPos,
        humanTiming,
        replayReadyTime[limb],
      );
      plan = applyRigEnvelope(plan);
      plan = appendRecoveryKeyframe(plan, recovery);
      motionPlans.push(plan);

      computeImpact(plan, ev.instrument, technique);

      fatigue = updateFatigue(fatigue, limb, technique, ev.timeSeconds);

      replayPosition[limb] = [ev.target.x, ev.target.y, ev.target.height + recovery.reboundHeightM];
      replayReadyTime[limb] = recovery.readyTimeS;
      replayLastTime[limb] = ev.timeSeconds;

      computeBodyState(ev.timeSeconds, replayPosition);

      const isDominantHand =
        (limb === 'RH' || limb === 'LH') && limb === `${effectiveStyle.dominantHand}H`;
      // Body-crossing is scored inside Rule 7's candidate search, but a
      // committed SequenceDecision does not carry that tag forward;
      // `false` is the safe default here, and the interface leaves room
      // to thread the real flag through in a future revision.
      this.calibrator.observe(isDominantHand, false);

      performanceEvents.push({
        eventId: ev.eventId,
        sourceId: ev.sourceId,
        timeSeconds: humanTiming.performedTimeS,
        limb,
        instrument: ev.instrument,
        strokeType: technique.strokeType,
        velocity: humanizedVelocity,
        microtimingOffsetMs: humanTiming.microtimingOffsetMs,
        dynamicLevel: technique.dynamicLevel,
        ruleTrace: ['rule08/29', 'rule09', 'rule10', 'rule12', 'rule13-19'],
      });
    }

    // ---- Assemble timeline (Rule 19) & sync (Rule 24) ----
    const timeline = assembleTimeline(motionPlans, techniqueByEvent);
    checkAudioAnimationSync(timeline);

    // ---- Validate & repair (Rule 25) ----
    const validation = validateAndRepair(performanceEvents);

    return {
      events: performanceEvents,
      animation: timeline,
      validation,
      durationS: timeline.durationS,
      seed: cfg.seed,
      engineVersion: '1.0.0',
    };
  }
}

// Imported separately so `engine.ts` reads as orchestration rather than
// as a wall of imports; `buildMotionPlan`'s own module owns the physics.
import { buildMotionPlan as buildMotionPlanFor } from './rule13-19-24-motion.js';
